### PR TITLE
10/Track enemy moves

### LINKED ISSUE #354: [Team 10] Tracks Enemy Moves
# **Description**
**Task:** [Tracks Enemy Moves ](#3).
**Feature:** [Feature which this tasks contributes to implementing](#2)

Special moves are unique abilities that enemies can use during combat, that is controlled by enemy ai which will make enemy moves predictable giving player an upper hand (if player predicts) else it will give enemy a tactical advantage in the combat by either disabling a player move button for next turn.
The feature tracks enemy actions during combat to enable special moves after completing a specific move combination. Each enemy type has a predefined sequence of actions that, when completed, triggers a special move, giving the enemy a tactical advantage by disabling the player's movement button for the next turn.

## **Example**

Let's take the Frog enemy as an example. The Frog can perform a special move to disable the player's movement button, but only after completing the move combo: Attack, Guard, Sleep.

Move 1 (Attack): The Frog starts by attacking. The move is stored in the EnemyMoveStore.

Current move sequence: [ATTACK]
Move 2 (Guard): On the next turn, the Frog guards. This action is added to the move list.

Current move sequence: [ATTACK, GUARD]
Move 3 (Sleep): The Frog then chooses to sleep. The EnemyMoveStore now contains three moves, triggering a combo check.

Current move sequence: [ATTACK, GUARD, SLEEP]
Combo Check: The system recognizes that the Frog has completed the combo "Attack, Guard, Sleep," so the Frog is now able to use its special move.

Special Move: The Frog’s special move activates, and it disables the player's movement button for the next turn, preventing the player from moving.

Reset: After the special move is executed, the EnemyMoveStore for the Frog is reset, allowing it to build up another combo for future special moves.
 
# **Dependencies**

- [ ] Enemy AI implementation
- [ ] Implementation of special effects

# **Milestones**
List of steps that need to be completed for this task.
- [ ] Implement move combo tracking logic for each enemy type( Sep. 18)
- [ ] Verifying special move combos during combat (Sep. 23).
- [ ] Implement the ability for enemies to use special moves once their combo is complete (Sep. 25).

**Completion Deadline:** Oct. 1

# **Documentation**
- [Main description of feature](../wiki/Special Moves)
- [JavaDoc](JavaDoc/Location/Package/Class)

# **Member**
- Sneha Zazen (sneha-zazen on GitHub) (zaza7_13 on Discord)

### PR DESCRIPTION
**Pull request by sneha-zazen**
# Description
This pull request implements the special moves feature for enemies in our combat system. It adds functionality to track enemy action sequences and trigger special moves when specific combinations are completed. This enhancement adds depth to enemy AI and creates more dynamic combat scenarios for players.

Closes #354 

# How Has This Been Tested?
The following tests were conducted using logger to verify the changes:

- [x] Test A: Verified that each enemy type correctly tracks its move sequence
- [x] Test B: Confirmed that special moves are triggered when the correct combination is achieved
- [x] Test C: Ensured that the specialmoveStore is properly reset after a special move is executed

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# Implementation Details
1. Added `UpdateSpecialMoveStore` method to track enemy actions.
2. Implemented `CheckSpecialMoveCombination` method to verify if a special move combination has been achieved.
3. Created switch cases for different enemy types (FROG, CHICKEN, MONKEY, BEAR) to handle their specific move combinations.
4. Added logging statements for better debugging and monitoring.

### CODE REVIEW

**Review by Howard-12 — APPROVED** [review]
looks good, has decent error handling.

File: source/core/src/main/com/csse3200/game/components/combat/CombatManager.java [hunk 1]
```diff
@@ -79,6 +83,8 @@ public CombatManager(Entity player, Entity enemy) {
         this.enemyMove = enemy.getComponent(CombatMoveComponent.class);
 
         this.moveChangedByConfusion = false;
+       // enemies = new HashMap<String,Entity>();
```
Comment (yessica-s) [inline source/core/src/main/com/csse3200/game/components/combat/CombatManager.java:86]: if you're not planning on using this you could remove it?

Comment (yessica-s) [inline source/core/src/main/com/csse3200/game/components/combat/CombatManager.java:86]: if you're not planning on using this you could remove it?

File: source/core/src/main/com/csse3200/game/components/combat/CombatManager.java [hunk 2]
```diff
@@ -240,8 +247,133 @@ private Action selectEnemyMove() {
...
+
+    private void CheckSpecialMoveCombination()
+    {
+        boolean NoSpecialMoveComboFlag= false;
+        ArrayList itemsList = EnemyMoveStore.get(enemy.getEnemyType().toString());
+        logger.info("Checking special move combination");
+        for (Map.Entry<String, ArrayList> entry : EnemyMoveStore.entrySet()) {
+
+    logger.info("Map<String,ArrayList> ::  " +entry.getKey()+ " :: "+entry.getValue() );
+}
```
Comment (yessica-s) [inline source/core/src/main/com/csse3200/game/components/combat/CombatManager.java:299]: could you fix the indenting of this bracket in the for loop? kinda confusing

**Review by ishusharma13 — APPROVED** [review]
good work!

**Comment by sneha-zazen** [comment]
resolved changes as per the review

**Pull request merged**